### ISSUE TITLE
Tracking issue for the consensus research

### ISSUE THREAD

**Issue by junha1** [postech-dao/simperby#4]

**Comment by byeongjee** [comment]
After discussion with @junha1 , we have come up with a slight variation of Tendermint suitable for our use cases. I'm looking forward to comments and suggestions.

We propose Calmdermint (the name is subject to change), an optimized version of Tendermint which supports early termination of voting phases before timeout expirations. Thanks to the early termination feature, a long timeout interval, possibly a week, can be adopted, while providing reasonable performance under practical conditions. Thus, nodes can be kept offline for the most of time, preventing unnecessary network congestion and blowup of the chain size, but can make progress as soon as sufficient number of them decide to make progress. The algorithm is guaranteed to be safe when the portion of byzantine nodes is less than 1/3, and it is also guaranteed to be sufficiently fast and make progress (under some fairness assumptions) when the portion of byzantine nodes is less than 1/9.

The following changes will be made to Tendermint consensus algorithm.
### AS-IS
- The prevote phase is finished if
	- (1a) the node sees prevote messages from more than 2/3 nodes, and
	- (1b) the timeout is exceeded
- The precommit phase is finished if
	- (2a) the node sees precommit messages from more than 2/3 nodes, and
	- (2b) the timeout is exceeded

### TO-BE
For the sake of simplicity, we assume there are total $9f + 1$ nodes where at most $f$ or them are byzantine.

- The prevote phase is finished if 
	- the conditions (1a) and (1b) are met, or
	- the node sees $8f + 1$ prevotes either to the proposal or to nil (early termination condition $a$)
- In the case of early termination of prevote,
	- If $7f + 1$ (or more) of the $8f+1$ prevotes are to the proposal, the node precommits to the proposal
	- Otherwise, the node precommits to nil
- The precommit phase is finished if
	-  the conditions (2a) and (2b) are met, or
	- the node sees $6f + 1$ precommits **to the proposal** (early termination condition $b$)


### Rough explanation for the early termination logic
Suppose that a proposal is broadcasted by the leader. Assuming all non-byzantine nodes are willing to prevote (either to the proposal or to nil) before timeout is expired, the early termination condition $a$ is met because ther are at least $8f + 1$  non-byzantine nodes. If the early timeout condition $a$ is met and $7f + 1$ of the collected prevotes are to the proposal, it is guaranteed that at least $6f +1$ non-byzantine nodes prevoted to the proposal, since the number of prevotes from byzantine nodes is at most $f$. The early termination condition $b$ should be met as well because there are $6f + 1$ non-byzantine nodes willing to precommit to the proposal.

The bottom line is that we have introduced the safe threshold for early termination
by weakening the fault threshold of the consensus, which is acceptable in our use cases.

**Comment by junha1** [comment]
Draft pseudo code
[vetomint-draft.pdf](https://github.com/postech-dao/simperby/files/9357282/vetomint-draft.pdf)

**Comment by byeongjee** [comment]
# Summary
After further revision, we come up with Vetomint, which is a variation of Tendermint consensus algorithm that supports preference-based voting on the consensus layer. Unlike Tendermint which requires validators to decide on a block based soley on its validity, participants of a Vetomint consensus system may choose to accept or reject a block based on their preferences. This allows flexibility on the choice of agendas that can be handled by the blockchain system. We have revised the early termination feature of the consensus algorithm to achieve the goal without performance loss, which is critical in our use case where the timeout interval is significant. Vetomint guarantees safety and liveness under less than 1/6 byzantine voting power and an assumption that innocent nodes will eventually agree on a proposal in a unanimous manner.


## Limitations of Tendermint
Although Tendermint adopts a voting-like procedure to achieve byzantine fault tolerance, it is not designed to support decision-making with preferences. That is, Tendermint requires a shared, deterministic oracle between all innocent participants to determine whether a proposal is acceptable or not, and the preference of each node has nothing to do with this process. This is the case for typical blockchain systems, where whether a proposal block is valid or not is determined soley by the current state of the blockchain and the content of the block. In such systems, validators play the role of "validator", not a participant of a democratic system.

However, in reality, we need votes in order to make a unified decision over multiple possible preferences, which cannot be done in Tendermint. For example, a presidential election has no notion of correctness; it is a matter of preference. 

Note that one may implement a governance protocol on top of the consensus layer, but it brings extra complexity to the system, especially in our simple use case where voters and validators are indistinguishable.

# Improvement
The main obstacle with using Tendermint for preference-based voting is that the prevote step of Tendermint may fall into a deadlock-like state, a stuck state that can be resolved only through waiting for a timeout expiration, if votes are split. For example, if half of the validators prevote to a proposal while the other half prevote to nil (i.e. they used veto), nothing but timeout rules can make progress. This is especially problematic for us because we want to adopt a long timeout interval, to reduce traffic and allow nodes to be kept offline for the most of time.

Augmenting the Tendermint protocol with an early termination rule requires lowering the byzantine threshold. This is because there must be a buffer threshold between 2/3 and 1 so that a node can early terminate a prevote step if it receives prevote messages exceeding the buffer threshold. Note that 2/3 should not be used for the threshold for early termination, because it allows a single byzantine node to disturb the consensus on an unanimously agreed proposal. We discovered that 5/6 is the desired threshold.

The pseudocode of the algorithm can be found here. 
[vetomint.pdf](https://github.com/postech-dao/simperby/files/9358491/consensus-standalone.pdf)
